### PR TITLE
[BUG] Disable old parent product when product simple moves to a configurable

### PR DESCRIPTION
**Pull request by lruozzi9**
Changes proposed in this pull request:
- When a simple product becomes part of a product model on Akeneo the current implementation will leave the old parent product without variants on Sylius. This will cause various errors. These changes will fix this kind of errors.

@webgriffe/wg-devs what do you think about this PR :shipit:?

### CODE REVIEW

File: src/Product/Importer.php [hunk 1]
```diff
@@ -354,4 +357,27 @@ public function reconcile(array $identifiersToReconcileWith): void
...
+        if ($oldParentProduct->getVariants()->count() !== 1) {
+            return;
+        }
+        $productVariant = $oldParentProduct->getVariants()->first();
+        if (!$productVariant instanceof ProductVariantInterface) {
+            return;
+        }
+        if ($productVariant->getCode() !== $identifier) {
+            return;
+        }
```
Comment (mmenozzi) [inline src/Product/Importer.php:379]: Il metodo si chiama `disableOldParentProductIfItHasNotAnyVariants` ma poi però lo si disabilita solo se ha 1 variante con quell'identificatore passato.

Come mai non avete preferito questa strada andando poi a spostare la chiamata del metodo dopo che la variante è stata associata al nuovo parent?

```suggestion
        if (!$oldParentProduct->getVariants()->isEmpty()) {
            return;
        }
```

Comment (lruozzi9) [reply src/Product/Importer.php:379]: Perché è vero che la variante viene associata al nuovo Product, ma nel vecchio Product la collezioni di varianti rimane invariata. Il metodo setProduct sulla ProductVariant dovrebbe essere così:
```php
  public function setProduct(?ProductInterface $product): void
  {
+    if ($this->product !== $product) {
+        $this->product->removeVariant($this);
+    }
      $this->product = $product;
  }
```

Comment (mmenozzi) [reply src/Product/Importer.php:379]: Ok

**Review by mmenozzi — APPROVED** [review]

File: src/Product/Importer.php [hunk 2]
```diff
@@ -354,4 +357,27 @@ public function reconcile(array $identifiersToReconcileWith): void
...
+
+    private function disableOldParentProductIfItHasNotAnyVariants(string $identifier, ProductInterface $product): void
+    {
+        $oldParentProduct = $this->productRepository->findOneByCode($identifier);
+        if ($oldParentProduct === null) {
+            return;
+        }
+        if ($oldParentProduct === $product) {
+            return;
+        }
```
Comment (LucaGallinari) [inline src/Product/Importer.php:369]: I'd have put this check in the import method as a wrapper for this method call. Something like this:
```
if ($oldParentProduct !== $product) {
    $this->disableOldParentProductIfItHasNotAnyVariants($identifier);
}
```
This way we'd have one less parameter and we call the method only if "necessary".

File: tests/Integration/DataFixtures/ORM/resources/ProductVariant/1111111186.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Sylius\Component\Core\Model\ProductVariant:
```
Comment (LucaGallinari) [inline tests/Integration/DataFixtures/ORM/resources/ProductVariant/1111111186.yaml:1]: This fixture doesn't seems to be used

File: src/Product/Importer.php (hunk 2 above)
Comment (lruozzi9) [reply src/Product/Importer.php:369]: We can remove the $product parameter but then we need to add the $oldParentProduct to check for the variants in it.

File: tests/Integration/DataFixtures/ORM/resources/ProductVariant/1111111186.yaml (hunk 3 above)
Comment (lruozzi9) [reply tests/Integration/DataFixtures/ORM/resources/ProductVariant/1111111186.yaml:1]: Done https://github.com/webgriffe/SyliusAkeneoPlugin/commit/328efd3840bf34d87235b61502fb7af600297f40

File: src/Product/Importer.php (hunk 2 above)
Comment (LucaGallinari) [reply src/Product/Importer.php:369]: Yes sorry, the $oldParentProduct variable is defined in this method.
We could do a check on the identifier/code instead
```
if ($identifier !== $product->getoCode()) {
    $this->disableOldParentProductIfItHasNotAnyVariants($identifier);
}
```
but only if it's not strictly necessary to do a check on the entire object ($oldParentProduct !== $product).

Comment (lruozzi9) [reply src/Product/Importer.php:369]: Yes, it should works https://github.com/webgriffe/SyliusAkeneoPlugin/commit/6a97504527f94b678f991667ea621c536f9052b9. Done!

Comment (LucaGallinari) [reply src/Product/Importer.php:369]: Great!

**Pull request merged**